File: algorithms/cpp/insertDeleteGetrandomO1.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <random>


using namespace std;


class RandomizedSet {
private:
    unordered_map<int, int> valToIndex;
    vector<int> values;
    mt19937 gen;
public:
    RandomizedSet() : gen(random_device{}()) {}
    
    bool insert(int val) {
        if (valToIndex.count(val)) { return false;}
        values.push_back(val);
        valToIndex[val] = values.size() - 1;
        return true;
    }
    
    bool remove(int val) {
        if (!valToIndex.count(val)) { return false;}
        int index = valToIndex.at(val);
        // Move the last element to the position of the element to remove
        values[index] = values[values.size() - 1];
        // Update the index of the element that was moved
        valToIndex[values[index]] = index;
        // Remove the last element from the vector
        values.pop_back();
        // Remove the element from the map
        valToIndex.erase(val);
        return true;
    }
    
    int getRandom() {
        uniform_int_distribution<int> dist(0, values.size() - 1);
        return values[dist(gen)];
    }
};

File: algorithms/python/insertDeleteGetrandomO1.py
import random


class RandomizedSet:
    def __init__(self):
        self.randomized_set = []
        self.hashmap = {}

    def insert(self, val: int) -> bool:
        if val in self.hashmap:
            return False
        self.randomized_set.append(val)
        self.hashmap[val] = len(self.randomized_set) - 1
        return True
        
    def remove(self, val: int) -> bool:
        if val not in self.hashmap: return False

        index = self.hashmap[val]
        
        self.hashmap[self.randomized_set[-1]] = index
        del self.hashmap[val]

        self.randomized_set[index] = self.randomized_set[-1]

        self.randomized_set.pop()
        return True

    def getRandom(self) -> int:
        return random.choice(self.randomized_set)
